### PR TITLE
allow components to be locked to block height

### PR DESCRIPTION
**Pull request by mpeterdev**
I branched off `0.3.0` for testing in alpha.near.org — will rebase against main branch once changes are conceptually approved

note that this is only an attempt at block height versioning for meeting basic needs. We should trend towards semver as a best practice but that will take more time to implement

~example usage: `<Widget src="mpmain2.near/widget/VersionTestInner" version="86775507" />`~
example usage: `<Widget src="mpmain2.near/widget/VersionTestInner@86775507" />`

### CODE REVIEW

File: src/lib/components/Widget.js [hunk 1]
```diff
@@ -112,11 +113,38 @@ export function Widget(props) {
...
+          "final"
+        );
+
+        // drill down in response to the version array
+        const srcTokens = src.split("/");
+        const componentVersions = srcTokens.reduce((acc, curr) => {
+          return acc[curr];
+        }, componentVersionRes);
+        // console.log("component versions", componentVersions);
+        if (!componentVersions.includes(version)) {
```
Comment (evgenykuzyakov) [inline src/lib/components/Widget.js:139]: I don't think this is quite necessary to check history, and this is also an expensive call. Probably the API can return a blockHeight at which the data was stored. So we can compare it this way if it's required.

**Review by evgenykuzyakov — COMMENTED** [review]
1. I'd prefer if we lock version using `@` sign as part of src. For example `a.near/widget/Widget@524534`.  This way it makes it easier to lock it within a direct url.
2. We should think about the default logic for children components. My original idea for block versions is that all children components will be automatically locked based on the height of the parent component. But it also might be undesirable in some situations. So we can add some modifiers on top of the height. E.g.
  - `@=524534`
  - `@^524534`
  - or something like this.

**Comment by mpeterdev** [comment]
> 1. I'd prefer if we lock version using `@` sign as part of src. For example `a.near/widget/Widget@524534`.  This way it makes it easier to lock it within a direct url. We can also use branches the same way

I equally considered going that way, happy to make the change

> 2. We should think about the default logic for children components. My original idea for block versions is that all children components will be automatically locked based on the height of the parent component. But it also might be undesirable in some situations. So we can add some modifiers on top of the height. E.g.
> 
> * `@=524534`
> * `@^524534`
> * or something like this.

Auto-locking is interesting. I would be inclined to push this change through such that it enables manual locking but doesn't affect global default behavior, then spin up a thread with more stakeholders to discuss changing default behavior as a later add-on.

For the example of `@^524534`, wouldn't that always just translate to `latest` without semver to define a max?

How about the following:
```
@<block>
e.g. @524534

@v<semver>
e.g. @v^5.0.0
```

alternatively for semver, do it npm-like
@semver:<semver>
e.g. @semver:^5.0.0

down the road, allow explicitly specifying no lock
@latest

if we decide to always require a version and want to support locking to parent, I could think of a number of options
@here @this @publish @parent
```

File: src/lib/components/Widget.js (hunk 1 above)
Comment (mpeterdev) [reply src/lib/components/Widget.js:139]: sounds good, I was leaning towards throwing an error if a matching version wasn't found to catch user error of specifying an incorrect block, but it doesn't need to be done here. If we care about that, we could do it during editing

**Comment by evgenykuzyakov** [comment]
I think this is a good start

**Comment by mpeterdev** [comment]
closed in favor of #10

**Pull request closed without merging**